Sources/PT.PM.Common/MessagePack/RootUstMessagePackSerializer.cs

```diff
@@ -69,52 +73,423 @@ public static byte[] Serialize(RootUst rootUst, bool isLineColumn, bool compress
+            var rootUstSerializer =  new RootUstMessagePackSerializer
             {
-                textSpanFormatter = textSpanFormatter,
-                fileFormatter = sourceFileFormatter,
-                rootUstFormatter = rootUstFormatter
+                serializedFile = serializedFile,
+                textSpanSerializer = textSpanSerializer,
+                fileSerializer = sourceFileSerializer
             };
 
             byte[] data2 = data ?? MessagePackUtils.UnpackDataIfRequired(serializedFile.Data);
 
-            RootUst result = readerResolver.GetFormatterWithVerify<RootUst>()
-                    .Deserialize(data2, 0, readerResolver, out readSize);
-            result.FillAscendants();
+            RootUst result = rootUstSerializer.DeserializeRootUst(data2, sourceFiles, 0, out readSize);
 
             return result;
         }
 
-        public IMessagePackFormatter<T> GetFormatter<T>()
+        private int SerializeRootUst(ref byte[] bytes, int offset, RootUst rootUst)
+        {
+            int newOffset = offset;
+
+            CurrentRoot = rootUst;
+
+            var localSourceFiles = new List<TextFile> {CurrentRoot.SourceFile};
+            rootUst.ApplyActionToDescendantsAndSelf(ust =>
+            {
+                foreach (TextSpan textSpan in ust.TextSpans)
+                {
+                    if (textSpan.File != null && !localSourceFiles.Contains(textSpan.File))
+                    {
+                        localSourceFiles.Add(textSpan.File);
+                    }
+                }
+            });
+
+            textSpanSerializer.LocalSourceFiles = localSourceFiles.ToArray();
+
+            //int sizeOffset = newOffset;
+            //newOffset += 5;
+
+            newOffset += MessagePackBinary.WriteBoolean(ref bytes, newOffset, textSpanSerializer.IsLineColumn);
+
+            newOffset += SerializeObject(ref bytes, newOffset, typeof(List<TextFile>), localSourceFiles);
+
+            newOffset += SerializeUst(ref bytes, newOffset, rootUst);
+
+            int writeSize = newOffset - offset;
+            //MessagePackBinary.WriteInt32ForceInt32Block(ref bytes, sizeOffset, writeSize - 5); TODO implement later
+
+            return writeSize;
+        }
+
+        private RootUst DeserializeRootUst(byte[] bytes, HashSet<IFile> sourceFiles, int offset, out int readSize)
+        {
+            int newOffset = offset;
+
+            textSpanSerializer.IsLineColumn = MessagePackBinary.ReadBoolean(bytes, newOffset, out int size);
+            newOffset += size;
+
+            var localSourceFiles = (TextFile[])DeserializeObject(bytes, newOffset, typeof(TextFile[]), out size);
+            newOffset += size;
+            textSpanSerializer.LocalSourceFiles = localSourceFiles;
+
+            lock (sourceFiles)
+            {
+                foreach (TextFile localSourceFile in localSourceFiles)
+                {
+                    sourceFiles.Add(localSourceFile);
```

Comment:
Are duplicates allowed in `sourceFiles`? Could there be duplicates if we unconditionally add files from different threads?

Reply:
I suppose that `localSourceFiles` collection will be used for both **Dfg** and **Ust** `TextFile` mappings, won't it?

Reply:
@flippaa, no, it is used only for **Ust**. Every **Dfg** is linked with **Ust**,  that's why there is no need to use this collection for **Dfg**.

Reply:
@gabriel-fallen , duplications are not existed, because the `HashSet` is used.